// lumapps-sdk-js/hooks/useContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useContext = void 0;
var react_redux_1 = require("react-redux");
var reselect_1 = require("reselect");
var config = null;
var get = function () {
    var _a, _b, _c, _d;
    var host = (_b = (_a = window.APPLICATION) === null || _a === void 0 ? void 0 : _a.HOST) !== null && _b !== void 0 ? _b : window.location.host;
    var protocol = (_d = (_c = window.APPLICATION) === null || _c === void 0 ? void 0 : _c.PROTOCOL) !== null && _d !== void 0 ? _d : window.location.protocol;
    config = {
        host: host,
        protocol: protocol,
        applicationHost: protocol + "//" + host,
    };
    return config;
};
var contentSelector = function (state) { return state === null || state === void 0 ? void 0 : state.content; };
var contentIdSelector = reselect_1.createSelector(contentSelector, function (content) { return content === null || content === void 0 ? void 0 : content.id; });
var instanceSelector = function (state) { return state === null || state === void 0 ? void 0 : state.instance; };
var instanceIdSelector = reselect_1.createSelector(instanceSelector, function (instance) { return instance === null || instance === void 0 ? void 0 : instance.id; });
var useContext = function () {
    var Config = get();
    return {
        contentId: react_redux_1.useSelector(contentIdSelector),
        instanceId: react_redux_1.useSelector(instanceIdSelector),
        baseUrl: Config.applicationHost,
    };
};
exports.useContext = useContext;
