// lumapps-sdk-js/hooks/useCurrentUser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useCurrentUser = void 0;
var react_redux_1 = require("react-redux");
var reselect_1 = require("reselect");
var userSliceSelector = function (state) { return state.user; };
var connectedUserSelector = reselect_1.createSelector(userSliceSelector, function (user) { return user.connectedUser; });
var apiProfileSelector = reselect_1.createSelector(connectedUserSelector, function (user) { return user === null || user === void 0 ? void 0 : user.apiProfile; });
var tokenSelector = reselect_1.createSelector(connectedUserSelector, function (user) { return user === null || user === void 0 ? void 0 : user.token; });
var emailSelector = reselect_1.createSelector(apiProfileSelector, function (apiProfile) { return apiProfile === null || apiProfile === void 0 ? void 0 : apiProfile.primaryEmail; });
var firstNameSelector = reselect_1.createSelector(apiProfileSelector, function (apiProfile) { var _a; return (_a = apiProfile === null || apiProfile === void 0 ? void 0 : apiProfile.name) === null || _a === void 0 ? void 0 : _a.givenName; });
var fullNameSelector = reselect_1.createSelector(apiProfileSelector, function (apiProfile) { var _a; return (_a = apiProfile === null || apiProfile === void 0 ? void 0 : apiProfile.name) === null || _a === void 0 ? void 0 : _a.fullName; });
var lastNameSelector = reselect_1.createSelector(apiProfileSelector, function (apiProfile) { var _a; return (_a = apiProfile === null || apiProfile === void 0 ? void 0 : apiProfile.name) === null || _a === void 0 ? void 0 : _a.familyName; });
var thumbnailPhotoUrlSelector = reselect_1.createSelector(connectedUserSelector, function (user) { return user === null || user === void 0 ? void 0 : user.profilePictureUrl; });
var useCurrentUser = function () {
    return {
        email: react_redux_1.useSelector(emailSelector),
        firstName: react_redux_1.useSelector(firstNameSelector),
        fullName: react_redux_1.useSelector(fullNameSelector),
        lastName: react_redux_1.useSelector(lastNameSelector),
        thumbnailPhotoUrl: react_redux_1.useSelector(thumbnailPhotoUrlSelector),
        token: react_redux_1.useSelector(tokenSelector),
    };
};
exports.useCurrentUser = useCurrentUser;
