// @lumx/react/esm/_internal/Avatar2.js
import { AspectRatio, Size, Theme } from './components.js';
import { g as getRootClassName, a as _objectWithoutProperties, b as _extends, d as classnames, h as handleBasicClasses } from './getRootClassName.js';
import React, { forwardRef } from 'react';
import { T as Thumbnail } from './Thumbnail2.js';

/**
 * Avatar sizes.
 */

/**
 * Component display name.
 */
var COMPONENT_NAME = 'Avatar';
/**
 * Component default class name and class prefix.
 */

var CLASSNAME = getRootClassName(COMPONENT_NAME);
/**
 * Component default props.
 */

var DEFAULT_PROPS = {
  size: Size.m,
  theme: Theme.light
};
/**
 * Avatar component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */

var Avatar = forwardRef(function (props, ref) {
  var actions = props.actions,
      alt = props.alt,
      badge = props.badge,
      className = props.className,
      image = props.image,
      onClick = props.onClick,
      onKeyPress = props.onKeyPress,
      size = props.size,
      theme = props.theme,
      thumbnailProps = props.thumbnailProps,
      forwardedProps = _objectWithoutProperties(props, ["actions", "alt", "badge", "className", "image", "onClick", "onKeyPress", "size", "theme", "thumbnailProps"]);

  return React.createElement("div", _extends({
    ref: ref
  }, forwardedProps, {
    className: classnames(className, handleBasicClasses({
      prefix: CLASSNAME,
      size: size,
      theme: theme
    }))
  }), React.createElement(Thumbnail, _extends({
    className: "".concat(CLASSNAME, "__thumbnail"),
    onClick: onClick,
    onKeyPress: onKeyPress
  }, thumbnailProps, {
    aspectRatio: AspectRatio.square,
    size: size,
    image: image,
    alt: alt
  })), actions && React.createElement("div", {
    className: "".concat(CLASSNAME, "__actions")
  }, actions), badge && React.createElement("div", {
    className: "".concat(CLASSNAME, "__badge")
  }, badge));
});
Avatar.displayName = COMPONENT_NAME;
Avatar.className = CLASSNAME;
Avatar.defaultProps = DEFAULT_PROPS;

export { Avatar as A };
