// @lumx/react/esm/_internal/Button2.js
import { Emphasis, Size, Theme } from './components.js';
import { g as getRootClassName, a as _objectWithoutProperties, d as classnames, e as getBasicClass, b as _extends, c as _objectSpread2 } from './getRootClassName.js';
import React, { forwardRef } from 'react';
import isEmpty from 'lodash/isEmpty';
import { B as ButtonRoot } from './ButtonRoot.js';
import { I as Icon } from './Icon2.js';

/**
 * Button emphasis definition.
 * @deprecated Use Emphasis instead.
 */

var ButtonEmphasis = Emphasis;
/**
 * Defines the props of the component.
 */

/**
 * Component display name.
 */
var COMPONENT_NAME = 'Button';
/**
 * Component default class name and class prefix.
 */

var CLASSNAME = getRootClassName(COMPONENT_NAME);
/**
 * Component default props.
 */

var DEFAULT_PROPS = {
  emphasis: Emphasis.high,
  size: Size.m,
  theme: Theme.light
};
/**
 * Button component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */

var Button = forwardRef(function (props, ref) {
  var children = props.children,
      className = props.className,
      emphasis = props.emphasis,
      leftIcon = props.leftIcon,
      rightIcon = props.rightIcon,
      size = props.size,
      theme = props.theme,
      forwardedProps = _objectWithoutProperties(props, ["children", "className", "emphasis", "leftIcon", "rightIcon", "size", "theme"]);

  var buttonClassName = classnames(className, getBasicClass({
    prefix: CLASSNAME,
    type: 'hasLeftIcon',
    value: !isEmpty(leftIcon)
  }), getBasicClass({
    prefix: CLASSNAME,
    type: 'hasRightIcon',
    value: !isEmpty(rightIcon)
  }));
  return React.createElement(ButtonRoot, _extends({
    ref: ref
  }, _objectSpread2({
    emphasis: emphasis,
    size: size,
    theme: theme
  }, forwardedProps), {
    className: buttonClassName,
    variant: "button"
  }), leftIcon && !isEmpty(leftIcon) && React.createElement(Icon, {
    icon: leftIcon
  }), children && React.createElement("span", null, children), rightIcon && !isEmpty(rightIcon) && React.createElement(Icon, {
    icon: rightIcon
  }));
});
Button.displayName = COMPONENT_NAME;
Button.className = CLASSNAME;
Button.defaultProps = DEFAULT_PROPS;

export { ButtonEmphasis as B, Button as a };
