// @lumx/react/esm/_internal/components.js
/**
 * Alignments.
 */
var Alignment = {
  bottom: 'bottom',
  center: 'center',
  end: 'end',
  left: 'left',
  right: 'right',
  spaceAround: 'space-around',
  spaceBetween: 'space-between',
  start: 'start',
  top: 'top'
};

/**
 * See SCSS variable $lumx-color-palette
 */
var ColorPalette = {
  primary: 'primary',
  secondary: 'secondary',
  blue: 'blue',
  dark: 'dark',
  green: 'green',
  yellow: 'yellow',
  red: 'red',
  light: 'light',
  grey: 'grey'
};

/**
 * See SCSS variable $lumx-color-variants
 */
var ColorVariant = {
  D1: 'D1',
  D2: 'D2',
  L1: 'L1',
  L2: 'L2',
  L3: 'L3',
  L4: 'L4',
  L5: 'L5',
  L6: 'L6',
  N: 'N'
};
var Theme = {
  light: 'light',
  dark: 'dark'
};
var Size = {
  xxs: 'xxs',
  xs: 'xs',
  s: 's',
  m: 'm',
  l: 'l',
  xl: 'xl',
  xxl: 'xxl',
  tiny: 'tiny',
  regular: 'regular',
  big: 'big',
  huge: 'huge'
};
var Orientation = {
  horizontal: 'horizontal',
  vertical: 'vertical'
};
var Emphasis = {
  low: 'low',
  medium: 'medium',
  high: 'high'
};

/**
 * List of typographies.
 */
var Typography = {
  overline: 'overline',
  caption: 'caption',
  body1: 'body1',
  body2: 'body2',
  subtitle1: 'subtitle1',
  subtitle2: 'subtitle2',
  title: 'title',
  headline: 'headline',
  display1: 'display1'
};

/**
 * All available aspect ratios.
 */
var AspectRatio = {
  /** Intrinsic content ratio. */
  original: 'original',

  /** Ratio 16:9 */
  wide: 'wide',

  /** Ratio 3:2 */
  horizontal: 'horizontal',

  /** Ratio 3:2 */
  vertical: 'vertical',

  /** Ratio 1:1 */
  square: 'square',

  /** Ratio constrained by the parent. */
  free: 'free'
};

/**
 * Semantic info about the purpose of the component
 */
var Kind = {
  info: 'info',
  success: 'success',
  warning: 'warning',
  error: 'error'
};

export { Alignment, AspectRatio, ColorPalette, ColorVariant, Emphasis, Kind, Orientation, Size, Theme, Typography };
