// @lumx/react/esm/_internal/GridItem.js
import { Orientation } from './components.js';
import { g as getRootClassName, a as _objectWithoutProperties, b as _extends, d as classnames, _ as _defineProperty, h as handleBasicClasses } from './getRootClassName.js';
import React, { forwardRef } from 'react';

/**
 * Component display name.
 */
var COMPONENT_NAME = 'Grid';
/**
 * Component default class name and class prefix.
 */

var CLASSNAME = getRootClassName(COMPONENT_NAME);
/**
 * Component default props.
 */

var DEFAULT_PROPS = {
  orientation: Orientation.horizontal,
  wrap: 'nowrap'
};
/**
 * Grid component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */

var Grid = forwardRef(function (props, ref) {
  var children = props.children,
      className = props.className,
      gutter = props.gutter,
      hAlign = props.hAlign,
      orientation = props.orientation,
      vAlign = props.vAlign,
      wrap = props.wrap,
      forwardedProps = _objectWithoutProperties(props, ["children", "className", "gutter", "hAlign", "orientation", "vAlign", "wrap"]);

  return React.createElement("div", _extends({
    ref: ref
  }, forwardedProps, {
    className: classnames(className, "".concat(CLASSNAME, "-container"), _defineProperty({}, "".concat(CLASSNAME, "--h-align-").concat(hAlign), hAlign), _defineProperty({}, "".concat(CLASSNAME, "--v-align-").concat(vAlign), vAlign), handleBasicClasses({
      prefix: CLASSNAME,
      orientation: orientation,
      wrap: wrap,
      gutter: gutter
    }))
  }), children);
});
Grid.displayName = COMPONENT_NAME;
Grid.className = CLASSNAME;
Grid.defaultProps = DEFAULT_PROPS;

/**
 * Component display name.
 */
var COMPONENT_NAME$1 = 'GridItem';
/**
 * Component default class name and class prefix.
 */

var CLASSNAME$1 = getRootClassName(COMPONENT_NAME$1);
/**
 * GridItem component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */

var GridItem = forwardRef(function (props, ref) {
  var children = props.children,
      className = props.className,
      width = props.width,
      align = props.align,
      order = props.order,
      forwardedProps = _objectWithoutProperties(props, ["children", "className", "width", "align", "order"]);

  return React.createElement("div", _extends({
    ref: ref
  }, forwardedProps, {
    className: classnames(className, handleBasicClasses({
      prefix: CLASSNAME$1,
      width: width,
      order: order,
      align: align
    }))
  }), children);
});
GridItem.displayName = COMPONENT_NAME$1;
GridItem.className = CLASSNAME$1;

export { Grid as G, GridItem as a };
