// @lumx/react/esm/_internal/InputLabel.js
import { Theme } from './components.js';
import { g as getRootClassName, a as _objectWithoutProperties, b as _extends, d as classnames, h as handleBasicClasses } from './getRootClassName.js';
import React, { forwardRef } from 'react';

/**
 * Defines the props of the component.
 */

/**
 * Component display name.
 */
var COMPONENT_NAME = 'InputLabel';
/**
 * Component default class name and class prefix.
 */

var CLASSNAME = getRootClassName(COMPONENT_NAME);
/**
 * Component default props.
 */

var DEFAULT_PROPS = {
  theme: Theme.light
};
/**
 * InputLabel component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */

var InputLabel = forwardRef(function (props, ref) {
  var children = props.children,
      className = props.className,
      htmlFor = props.htmlFor,
      isRequired = props.isRequired,
      theme = props.theme,
      forwardedProps = _objectWithoutProperties(props, ["children", "className", "htmlFor", "isRequired", "theme"]);

  return React.createElement("label", _extends({
    ref: ref
  }, forwardedProps, {
    htmlFor: htmlFor,
    className: classnames(className, handleBasicClasses({
      prefix: CLASSNAME,
      isRequired: isRequired,
      theme: theme
    }))
  }), children);
});
InputLabel.displayName = COMPONENT_NAME;
InputLabel.className = CLASSNAME;
InputLabel.defaultProps = DEFAULT_PROPS;

export { InputLabel as I };
