// @lumx/react/esm/_internal/Message2.js
import { Size, ColorPalette, Kind } from './components.js';
import { g as getRootClassName, _ as _defineProperty, a as _objectWithoutProperties, b as _extends, d as classnames, h as handleBasicClasses } from './getRootClassName.js';
import React, { forwardRef } from 'react';
import { m as mdiAlert, a as mdiInformation, b as mdiCheckCircle, c as mdiAlertCircle } from './mdi.js';
import { I as Icon } from './Icon2.js';

var _CONFIG;
/**
 * Defines the props of the component.
 */

/**
 * Component display name.
 */
var COMPONENT_NAME = 'Message';
/**
 * Component default class name and class prefix.
 */

var CLASSNAME = getRootClassName(COMPONENT_NAME);
/**
 * Associative map from message kind to color and icon.
 */

var CONFIG = (_CONFIG = {}, _defineProperty(_CONFIG, Kind.error, {
  color: ColorPalette.red,
  icon: mdiAlert
}), _defineProperty(_CONFIG, Kind.info, {
  color: ColorPalette.dark,
  icon: mdiInformation
}), _defineProperty(_CONFIG, Kind.success, {
  color: ColorPalette.green,
  icon: mdiCheckCircle
}), _defineProperty(_CONFIG, Kind.warning, {
  color: ColorPalette.yellow,
  icon: mdiAlertCircle
}), _CONFIG);
/**
 * Message component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */

var Message = forwardRef(function (props, ref) {
  var children = props.children,
      className = props.className,
      hasBackground = props.hasBackground,
      kind = props.kind,
      customIcon = props.icon,
      forwardedProps = _objectWithoutProperties(props, ["children", "className", "hasBackground", "kind", "icon"]);

  var _ref = CONFIG[kind] || {},
      color = _ref.color,
      icon = _ref.icon;

  return React.createElement("div", _extends({
    ref: ref,
    className: classnames(className, handleBasicClasses({
      color: color,
      hasBackground: hasBackground,
      prefix: CLASSNAME
    }))
  }, forwardedProps), (customIcon || icon) && React.createElement(Icon, {
    className: "lumx-message__icon",
    icon: customIcon || icon,
    size: Size.xs
  }), React.createElement("div", {
    className: "lumx-message__text"
  }, children));
});
Message.displayName = COMPONENT_NAME;
Message.className = CLASSNAME;

export { Message as M };
