// @lumx/react/esm/_internal/Progress2.js
import { Theme } from './components.js';
import { g as getRootClassName, a as _objectWithoutProperties, b as _extends, d as classnames, h as handleBasicClasses } from './getRootClassName.js';
import React, { forwardRef } from 'react';

/**
 * Progress variants.
 */

var ProgressVariant = {
  linear: 'linear',
  circular: 'circular'
};

/**
 * Component display name.
 */
var COMPONENT_NAME = 'Progress';
/**
 * Component default class name and class prefix.
 */

var CLASSNAME = getRootClassName(COMPONENT_NAME);
/**
 * Component default props.
 */

var DEFAULT_PROPS = {
  theme: Theme.light,
  variant: ProgressVariant.circular
};
/**
 * Progress component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */

var Progress = forwardRef(function (props, ref) {
  var className = props.className,
      theme = props.theme,
      variant = props.variant,
      forwardedProps = _objectWithoutProperties(props, ["className", "theme", "variant"]);

  return React.createElement("div", _extends({
    ref: ref
  }, forwardedProps, {
    className: classnames(className, handleBasicClasses({
      prefix: CLASSNAME,
      theme: theme,
      variant: variant
    }))
  }), React.createElement("div", {
    className: classnames("".concat(CLASSNAME, "-").concat(variant))
  }, variant === ProgressVariant.circular && React.createElement(React.Fragment, null, React.createElement("div", {
    className: "lumx-progress-circular__double-bounce1"
  }), React.createElement("div", {
    className: "lumx-progress-circular__double-bounce2"
  }), React.createElement("svg", {
    className: "lumx-progress-circular__svg",
    viewBox: "25 25 50 50"
  }, React.createElement("circle", {
    className: "lumx-progress-circular__path",
    cx: "50",
    cy: "50",
    r: "20",
    fill: "none",
    strokeWidth: "5"
  }))), variant === ProgressVariant.linear && React.createElement(React.Fragment, null, React.createElement("div", {
    className: "lumx-progress-linear__line1"
  }), React.createElement("div", {
    className: "lumx-progress-linear__line2"
  }))));
});
Progress.displayName = COMPONENT_NAME;
Progress.className = CLASSNAME;
Progress.defaultProps = DEFAULT_PROPS;

export { ProgressVariant as P, Progress as a };
