// @lumx/react/esm/_internal/RadioGroup.js
import { Theme } from './components.js';
import { g as getRootClassName, a as _objectWithoutProperties, b as _extends, d as classnames, h as handleBasicClasses } from './getRootClassName.js';
import React, { forwardRef, useMemo } from 'react';
import { u as uid } from '../index2.js';
import { I as InputHelper } from './InputHelper.js';
import { I as InputLabel } from './InputLabel.js';

/**
 * Defines the props of the component.
 */

/**
 * Component display name.
 */
var COMPONENT_NAME = 'RadioButton';
/**
 * Component default class name and class prefix.
 */

var CLASSNAME = getRootClassName(COMPONENT_NAME);
/**
 * Component default props.
 */

var DEFAULT_PROPS = {
  theme: Theme.light
};
/**
 * RadioButton component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */

var RadioButton = forwardRef(function (props, ref) {
  var checked = props.checked,
      className = props.className,
      disabled = props.disabled,
      helper = props.helper,
      id = props.id,
      _props$isChecked = props.isChecked,
      isChecked = _props$isChecked === void 0 ? checked : _props$isChecked,
      _props$isDisabled = props.isDisabled,
      isDisabled = _props$isDisabled === void 0 ? disabled : _props$isDisabled,
      label = props.label,
      name = props.name,
      onChange = props.onChange,
      theme = props.theme,
      value = props.value,
      forwardedProps = _objectWithoutProperties(props, ["checked", "className", "disabled", "helper", "id", "isChecked", "isDisabled", "label", "name", "onChange", "theme", "value"]);

  var radioButtonId = useMemo(function () {
    return id || "".concat(CLASSNAME.toLowerCase(), "-").concat(uid());
  }, [id]);

  var handleChange = function handleChange(event) {
    if (onChange) {
      onChange(value, name, event);
    }
  };

  return React.createElement("div", _extends({
    ref: ref
  }, forwardedProps, {
    className: classnames(className, handleBasicClasses({
      isChecked: isChecked,
      isDisabled: isDisabled,
      isUnchecked: !isChecked,
      prefix: CLASSNAME,
      theme: theme
    }))
  }), React.createElement("div", {
    className: "".concat(CLASSNAME, "__input-wrapper")
  }, React.createElement("input", {
    className: "".concat(CLASSNAME, "__input-native"),
    disabled: isDisabled,
    id: radioButtonId,
    tabIndex: isDisabled ? -1 : 0,
    type: "radio",
    name: name,
    value: value,
    checked: isChecked,
    onChange: handleChange
  }), React.createElement("div", {
    className: "".concat(CLASSNAME, "__input-placeholder")
  }, React.createElement("div", {
    className: "".concat(CLASSNAME, "__input-background")
  }), React.createElement("div", {
    className: "".concat(CLASSNAME, "__input-indicator")
  }))), React.createElement("div", {
    className: "".concat(CLASSNAME, "__content")
  }, label && React.createElement(InputLabel, {
    htmlFor: radioButtonId,
    theme: theme,
    className: "".concat(CLASSNAME, "__label")
  }, label), helper && React.createElement(InputHelper, {
    theme: theme,
    className: "".concat(CLASSNAME, "__helper")
  }, helper)));
});
RadioButton.displayName = COMPONENT_NAME;
RadioButton.className = CLASSNAME;
RadioButton.defaultProps = DEFAULT_PROPS;

/**
 * Defines the props of the component.
 */

/**
 * Component display name.
 */
var COMPONENT_NAME$1 = 'RadioGroup';
/**
 * Component default class name and class prefix.
 */

var CLASSNAME$1 = getRootClassName(COMPONENT_NAME$1);
/**
 * RadioGroup component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */

var RadioGroup = forwardRef(function (props, ref) {
  var children = props.children,
      className = props.className,
      forwardedProps = _objectWithoutProperties(props, ["children", "className"]);

  return React.createElement("div", _extends({
    ref: ref
  }, forwardedProps, {
    className: classnames(className, handleBasicClasses({
      prefix: CLASSNAME$1
    }))
  }), children);
});
RadioGroup.displayName = COMPONENT_NAME$1;
RadioGroup.className = CLASSNAME$1;

export { RadioButton as R, RadioGroup as a };
